swift/README.md

```diff
@@ -1,372 +1,1100 @@
 
-Use the `weak` modifier for `delegate` properties.  
+```swift
+if let textContainer = self.textContainer {
+  // do many things with textContainer
+}
+```
 
-Name the delegate property `delegate` unless your class supports multiple different delegates.  
+When naming optional variables and properties, avoid naming them like `optionalString` or `maybeView` since their optional-ness is already in the type declaration.
 
-Protocols (aka Interfaces)
-==========================
+For optional binding, shadow the original name when appropriate rather than using names like `unwrappedView` or `actualLabel`.
 
-Protocols are awesome. Use them to restrict the API surface area of an object
-being passed around.
+Use `as` for type coercion if possible (this is enforced statically). Otherwise use `as?`.
 
-typealiases
-===========
+**Preferred:**
+```swift
+var subview: UIView?
+var volume: Double?
 
-`typealias` standard types if used in specific contexts.
+// later on...
+if let subview = subview, volume = volume {
+  // do something with unwrapped subview and volume
+}
+```
 
+**Not Preferred:**
 ```swift
-typealias MessageAddress = String
+var optionalSubview: UIView?
+var volume: Double?
+
+if let unwrappedSubview = optionalSubview {
+  if let realVolume = volume {
+    // do something with unwrappedSubview and realVolume
+  }
+}
 ```
 
-Also, you will often find it useful to typealias function types:
+### Struct Initializers
 
+Use the native Swift struct initializers rather than the legacy CGGeometry constructors.
+
+**Preferred:**
 ```swift
-typealias RpcMethodCallback = (RpcMethodResult) -> Void
+let bounds = CGRect(x: 40, y: 20, width: 120, height: 80)
+let centerPoint = CGPoint(x: 96, y: 42)
+```
+
+**Not Preferred:**
+```swift
+let bounds = CGRectMake(40, 20, 120, 80)
+let centerPoint = CGPointMake(96, 42)
 ```
 
-Access Control
-==============
+Prefer the struct-scope constants `CGRect.infinite`, `CGRect.null`, etc. over global constants `CGRectInfinite`, `CGRectNull`, etc. For existing variables, you can use the shorter `.zero`.
+
+
+### Lazy Initialization
 
-Use either `private` or nothing (which defaults to module/framework internal).
-Only use `public` once you've written a feature and need to expose it outside
-of a module. Most things do not need to be exposed.  Let the compiler guide
-you on what to make `public`.
+Consider using lazy initialization for finer grain control over object lifetime. This is especially true for `UIViewController` that loads views lazily. Use a closure that is immediately called `{ }()`. Example:
+
+```swift
+lazy var locationManager: CLLocationManager = {
+  let manager = CLLocationManager()
+  manager.desiredAccuracy = kCLLocationAccuracyBest
+  manager.delegate = self
+  manager.requestAlwaysAuthorization()
+  return manager
+}()
+```
 
-Comments
-========
+**Notes:**
+  - `[unowned self]` is not required here. A retain cycle is not created.
```

Comment:
Sure. Technically a retain cycle is created (but its super short-lived, so I don't count it). The reason it doesn't create a permanent retain cycle is because (as you say) the closure is executed once _and_ because the value assigned to the lazy var is the manager. A simple way to make a retain cycle here would be to change this to be:

```swift
lazy var locationManager: () -> CLLocationManager = { // [unowned self] in <- would break the retain cycle here
   let manager = CLLocationManager()
   manager.desiredAccuracy = kCLLocationAccuracyBest
   manager.delegate = self
   manager.requestAlwaysAuthorization()
   return manager
 }
```
this would create a permanent retain cycle because it's the closure that's being set to the lazy var, not the value that the closure returns, and the closure captures `self`. I'll add an explanation to the docs.